opendkim/opendkim-lua.3
.TH opendkim-lua 3 "The OpenDKIM Project"
.SH NAME
.B OpenDKIM Lua programming
- programming the OpenDKIM filter using Lua scripts
.SH DESCRIPTION
The OpenDKIM filter has hooks to run user-provided scripts for making policy
decisions regarding signatures to add on outbound messages or verification and
acceptance of messages inbound.  The hooks take the form of multiple Lua
scripts which, if defined, are run at important points during processing of 
a message.

For a full description of the Lua language, consult Lua programming references
(see below for a starting point).  This man page only describes the use of
Lua in the context of OpenDKIM, specifically the functions and global variables
OpenDKIM provides for use in user-constructed scripts beyond what Lua provides
by default.

Three scripting hooks are provided.  They are as follows:
.TP
.I setup
The setup script is run after all headers for the message have been received
but before any DKIM operations have started.  At this point the user can
examine the available header fields to decide whether the message should be
signed or verified (or both) and, if signing, which key(s) should be used to
add signatures and which signature features are desired.
.TP
.I screen
The screen script is run after the DKIM verification context has been
established.  The main purpose of this script is to give the user an
opportunity to examine the message header fields compared to the available
DKIM signatures and determine which, if any, should be ignored during
verification.  For example, the user might decide only signatures added by
domains exactly matching that in the From: domain are acceptable, and the
rest should be ignored.
.TP
.I final
The final script is run after all of the DKIM verification and signing work
has been completed.  The user has an opportunity to examine the results
of all of the signature evaluations and make a decision about whether or not
the message should be accepted, rejected, discarded, quarantined, etc.
If the message is accepted, any signatures requested earlier will be
added to the messages before it is released.

Unless otherwise noted, all functions return a single result; on success
they return the requested data, and on error they return the Lua constant
"nil".
.SH GLOBAL VARIABLES
The following global variable(s) are provided for all user scripts:
.TP
.I ctx
This is a generic context pointer referring to the context in which the
filtering operation is being performed.  It represents a single message
in progress, and the connection that accepted it.
.SH SETUP SCRIPT FUNCTIONS
These functions are made available to Lua for processing a message through
the setup script:
.TP
.B odkim_check_popauth(ctx)
Returns 1 if the SMTP client represented by
.I ctx
is coming from an IP address found in the POPAUTH database (if enabled
and configured), and 0 otherwise.  Returns the Lua constant "nil" if
the POPAUTH database is not enabled or not configured.
.TP
.B odkim_get_clienthost(ctx)
Returns the name of the host on the other end of the SMTP connection
sending the current message.  This is usually a hostname, but might be
an IP address in square brackets if the SMTP client's IP address does not
have a name associated with it.
.TP
.B odkim_db_check(db, string)
Returns 1 if
.I db
refers to a valid database handle (see
.B odkim_get_dbhandle()
below) and
.I string
is found in that database, and 0 otherwise.  If an error occurs, the Lua
constant "nil" is returned.
.TP
.B odkim_get_dbhandle(ctx, db)
Returns a handle for the requested database that can be used in later
queries.  The value of
.I db
should be one of
.I DB_MTAS
(database of MTA names whose mail should be signed),
.I DB_MACROS
(database of MTA macro checks to be done when determining signing),
.I DB_DOMAINS
(database of domains to be signed),
.I DB_THIRDPARTY
(database of third party signatures to be trusted) and
.I DB_DONTSIGNTO
(database of recipients whose mail should not be signed).  If the requested
database is not set in the current configuration file, a Lua "nil" is
returned.
.TP
.B odkim_get_fromdomain(ctx)
Retrieves the domain name of the sender of the message represented by
.I ctx.
.TP
.B odkim_get_header(ctx, name, n)
Retrieves the string contained in instance
.I n
of the header field called
.I name
from the message represented by
.I ctx,
or the Lua constant "nil" if there was no such header field.
Header field numbering starts at 0, so use 0 for the first instance,
1 for the second, etc.  For example:

fromaddr = odkim_get_header(ctx, "From", 0)

This will return the value of the first (and hopefully only) "From" header
field.
.TP
.B odkim_get_mtasymbol(ctx, name)
Retrieves the value of the symbol called
.I name
from the MTA connection represented by
.I ctx,
or the Lua constant "nil" if the requested symbol was not available at the
time of the request.
.TP
.B odkim_get_rcpt(ctx, n)
Returns the
.I nth
envelope recipient for the message represented by
.I ctx.
Recipient numbering starts at 0, so for the first recipient, use 0 for
.I n.
If
.I n
references an out-of-range value, the Lua constant "nil" is returned.
.TP
.B odkim_internal_ip(ctx)
Returns 1 if the SMTP client is coming from an internal IP address, and 0
otherwise.
.TP
.B odkim_rcpt_count(ctx)
Returns the count of envelope recipients on the message.
.TP
.B odkim_resign(ctx)
Arranges that the arriving message will be verified and then re-signed in
a single operation.  Returns 1 on success or the Lua constant "nil" on failure.
.TP
.B odkim_sign(ctx, domain, selector)
Requests that the filter sign the message represented by
.I ctx
using the key whose
.I domain
and
.I selector
from the key list match the provided parameters.  See the
.B KeyList
parameter in
.I opendkim.conf(5)
for more information.  Returns 1 on success and 0 on failure.
.TP
.B odkim_use_ltag(ctx)
Requests that all signatures added to the message represented by
.I ctx
include "l=" (body length) tags.  Always returns the Lua constant "nil".
.TP
.B odkim_verify(ctx)
Requests that the message represented by
.I ctx
be subjected to DKIM signature verification.  Returns the Lua constant
"nil" on success, or an error string on failure.
.SH SCREEN SCRIPT FUNCTIONS
The screen script has the following functions available to it, whose
descriptions can be found above:
.B odkim_db_check,
.B odkim_fromdomain,
.B odkim_get_dbhandle,
.B odkim_get_header,
.B odkim_get_rcpt
and
.B odkim_rcpt_count.

The following additional functions are provided for this script:
.TP
.B odkim_get_sigcount(ctx)
Returns the number of signatures found in the message represented by
.I ctx,
or the Lua constant "nil" in case of an error.
.TP
.B odkim_get_sighandle(ctx, n)
Returns a handle representing an internal copy of the
.I nth
signature found on the message represented by
.I ctx.
.I n
must be a number greater than or equal to zero (representing the first
signature) and less than the number of signatures on the message, which
can be determined using
.B odkim_get_sigcount
above.  The requested handle is returned on success, or the Lua constant
"nil" is returned on failure.
.TP
.B odkim_sig_getdomain(sig)
Returns the name of the domain in the signature handle specified by
.I sig,
previously returned by a call to
.B odkim_get_sighandle().
This is taken from the signature's "d=" tag.
.TP
.B odkim_sig_getidentity(sig)
Returns the identity of the agent adding the signature handle specified by
.I sig,
previously returned by a call to
.B odkim_get_sighandle().
This is taken from the signature's "i=" tag.  This may be a default value
and not one that was explicitly part of the signature.  If the identity could
not be determined, the Lua constant "nil" is returned.
.TP
.B odkim_sig_ignore(sig)
Instructs the verification code to ignore completely the signature specified
by
.I sig,
previously returned by a call to
.B odkim_get_sighandle().
Any pending verification of the message will act as if that signature was
not present on the message.  Always returns the Lua constant "nil".
.SH FINAL SCRIPT FUNCTIONS
The screen script has the following functions available to it, whose
descriptions can be found above:
.B odkim_get_rcpt,
.B odkim_get_sigcount,
.B odkim_get_sighandle,
.B odkim_sig_getdomain,
.B odkim_sig_getidentity,
and
.B odkim_rcpt_count.

The following additional functions are provided for this script:
.TP
.B odkim_add_rcpt(ctx, addr)
Adds
.I addr
as an envelope recipient to the message represented by
.I ctx.
Returns 1 on success, or the Lua constant "nil" on failure.
.TP
.B odkim_del_rcpt(ctx, addr)
Deletes
.I addr
from the list of envelope recipients on the message represented by
.I ctx,
and adds a new X-Original-Recipient: header field containing the deleted
address.  Returns 1 on success, or the Lua constant "nil" on failure.
.TP
.B odkim_get_policy(ctx)
Returns the discovered Author Domain Signing Practices (ADSP) policy, if one
was published and could be retrieved.  The value will be one of
.I DKIM_POLICY_UNKNOWN,
.I DKIM_POLICY_ALL
and
.I DKIM_POLICY_DISCARDABLE,
or the Lua constant "nil" if no policy could be retrieved.
.TP
.B odkim_get_presult(ctx)
Returns the result code from an attempt to retrieve the ADSP for the author
domain.  The value will be one of
.I DKIM_PRESULT_NXDOMAIN
(the sender's domain does not exist),
or
.I DKIM_PRESULT_AUTHOR
(an ADSP record was retrieved), or the Lua constnat "nil" if no policy
could be retrieved.
.TP
.B odkim_get_reputation(ctx, sig, qroot)
Performs a reputation query about the message represented by
.I ctx
and the signature indicated by
.I sig,
the latter returned by a previous call to
.B odkim_get_sighandle().
The
.I qroot
parameter is a string naming the DNS root of the published reputation
query data, or can be the empty string to use the built-in default.
The integer result is returned on success, or "nil" on error.
.TP
.B odkim_quarantine(ctx, reason)
Asks the MTA to quarantine the message represented by
.I ctx
using
.I reason
as a text string indicating the reason for the request.  Returns 1 on
success or the Lua constant "nil" on failure.
.TP
.B odkim_set_reply(ctx, rcode, xcode, message)
Instructs the MTA to return the specified SMTP reply to the client sending
the message represented by
.I ctx.
.I rcode
must be a three-digit SMTP reply code starting with 4 or 5 (for temporary
or permanent failures, respectively);
.I xcode
must be the empty string or a valid extended reply code (see RFC2034) matching
.I rcode;
and
.I message
must be the text portion of the SMTP reply to be sent.  Returns 1 on success
or the Lua constant "nil" on failure.
.TP
.B odkim_set_result(ctx, result)
Arranges to have the MTA return a specific result code in response
to the message represented by
.I ctx.
The value of
.I result
must be one of
.I SMFIS_TEMPFAIL
(temporary failure/rejection),
.I SMFIS_DISCARD
(accept but discard the message)
and
.I SMFIS_REJECT
(permanent failure/rejection).  Returns 1 on success or the Lua constant
"nil" on failure.
.TP
.B odkim_sig_bhresult(sig)
Returns the result code corresponding to the body hash evaluation of
the signature handled specified by
.I sig,
previously returned by a call to
.B odkim_get_sighandle().
Valid values are defined in the
.B libopendkim
header file
.I dkim.h
with DKIM_SIGBH_ prefixes, and will be made available by name in a
future version.
.TP
.B odkim_sig_bodylength(sig)
Returns the total length of the message signed by
.I sig,
previously returned by a call to
.B odkim_get_sighandle(),
or the Lua constant "nil" if this value could not be determined.
.TP
.B odkim_sig_canonlength(sig)
Returns the canonicalized length of the message signed by
.I sig,
previously returned by a call to
.B odkim_get_sighandle(),
or the Lua constant "nil" if this value could not be determined.
Note that this may be less than the value returned by
.B odkim_get_bodylength()
if the signature only covered part of the message.
.TP
.B odkim_sig_result(sig)
Returns the result code corresponding to the signature handled specified by
.I sig,
previously returned by a call to
.B odkim_get_sighandle().
Valid values are defined in the
.B libopendkim
header file
.I dkim.h
with DKIM_SIGERROR_ prefixes, and will be made available by name in a
future version.
.SH NOTES
TBD
.SH VERSION
This man page covers version 1.2.0 of
.I OpenDKIM.
.SH COPYRIGHT
Copyright (c) 2009, The OpenDKIM Project.  All rights reserved.
.SH SEE ALSO
.I opendkim(8),
.I opendkim.conf(5)
.P
Lua -- http://www.lua.org


RELEASE_NOTES
			OPENDKIM RELEASE NOTES
	$Id: RELEASE_NOTES,v 1.64 2009/12/01 23:18:08 cm-msk Exp $


This listing shows the versions of the OpenDKIM package, the date of
release, and a summary of the changes in that release.

1.2.0		2009/12/14
	Feature request #SF2873902: Overhaul the database backend code so that
		features that use external files or databases can be in any
		of several supported formats.  This will make adding new
		external data sources and formats simpler, and obviates the
		need for a bunch of individual feature requests.  This may
		have a few backward compatibility issues with respect to
		the configuration file.  Feature requested by Daniel Black.
	Feature request #SF2873900: Add optional support for OpenDBX for
		connecting to ODBC and SQL backend databases.  Requested
		by Daniel Black.
	Add "On-PolicyError" setting, allowing continuation of processing
		when an ADSP query fails.
	Activate _FFR_MULTIPLE_SIGNATURES, allowing optional addition of
		more than one signature per message passing through the
		filter.
	Add _FFR_RESIGN which allows binding of a signing handle to a
		verifying handle so that only one body hash needs to be
		run when a message will be re-signed as-is.  Suggested
		by Daniel Black.
	LIBOPENDKIM: Move VBR functions from libopendkim into their own
		new library, libvbr.
	LIBOPENDKIM: Rename the rfc2822_*() parsing functions to have
		"dkim_" prefixes, and rename their containing file
		accordingly.
	LIBOPENDKIM: Fixes in relaxed body canonicalization and chunk
		processing.  Problems noted by Masumi Taketomi Parekh
		of Yahoo!.
	LIBOPENDKIM: New library flag DKIM_LIBFLAGS_BADSIGHANDLES which
		asks the library to tolerate signature syntax errors and
		make such signatures available for limited inspection
		rather than completely ignoring them.  Requested by Masumi
		Taketomi Parekh of Yahoo!.
	BUILD: Split up library assignments between libopendkim and opendkim,
		taking advantage of libtool.  Based on a patch by Daniel Black.
	BUILD: When possible, limit the symbols exported as part of
		libopendkim to only those listed in dkim.h.
	BUILD: Generate opendkim.conf.5 man page containing all features,
		including FFR, annotated with if they are included and their 
		experimental status (for FFRs).
	BUILD: Improve static linking against openssl as noted by Roman
		Gelfand.  Compiles against openssl version 1.0.0-beta4 now. 
	BUILD: Provide pkg-config files {opendkim,vbr,ar}.pc for use by 
		other applications.
	BUILD: Fix up libresolv detection.
	BUILD: Add pkg-config checks for openssl, tre and opendbx packages to
		determine their installed library locations. Automate 
		versioning and deployment.
	BUILD: Added m4 macro library directory with updated ax_pthread.m4.
	BUILD: Moved feature, _FFR and library #defines from Makefiles to 
		build-config.h. Added macros for FFR and FEATURES.
	BUILD: Add support for versions of libtre older than 0.8.0.
	BUILD: Move all libopendkim tests into their own subdirectory.

1.1.2		2009/11/01
	Under _FFR_SENDER_MACRO, need to check the value of "SenderMacro"
		in the configuration file.  Noted by Daniel Black.
	Feature request #SF2873901: Add _FFR_REDIRECT which optionally
		redirects messages that fail verification to a specific
		address, storing the original recipients in
		X-Original-Recipient: header fields.  Suggested by
		Daniel Black.
	LIBOPENDKIM: Have dkim_dns_set_callback() return
		DKIM_STAT_NOTIMPLEMENT if the underlying resolver doesn't
		have a callback facility.  Suggested by Daniel Black.
	LIBOPENDKIM: Move internal-only types and macros from dkim.h
		to dkim-internal.h.  Based on an idea from Daniel Black.
	LIBOPENDKIM: Add in all previously optional functions so that the
		API is invariant regardless of selected features.  Add
		dkim_libfeature() as a way to detect availability of features
		at runtime.  Suggested by Daniel Black.
	LIBOPENDKIM: Some global namespace consolidation.  Suggested by
		Daniel Black.
	BUILD: Fix bug #SF2882206, patch #SF2880986: Handle libtre
		installations where "--enable-system-abi" was selected.
		Reported by Stevan Bajic; patch from Daniel Black.
	BUILD: Convert libar compilation to the libtool method.
		Suggested by Daniel Black.
	BUILD: Minor autoconf fixes, contributed by Daniel Black.

1.1.1		2009/10/09
	Plug a number of potential but minor memory and file handle leaks,
		remove some dead code, guard against NULL dereferences, fix an
		errant return code check, fix a double-free(), and several
		other fixes found by a code analysis tool, used by courtesy
		of Cloudmark.
	Change "x-dkim-adsp" to "dkim-adsp" in Authentication-Results header
		field code now that RFC5617 is published.
	Apply "On-DNSError" setting when policy queries fail.
	BUILD: Some compilation fixes and type cleanup.  Based on patches
		provided by Daniel Black.
	Add _FFR_SENDER_MACRO: Determine the message sender based on the
		contents of a macro instead of on a header field.
		Based on a patch from Ondrej Sury.

1.1.0		2009/09/16
	Feature request #SF2839110: Add _FFR_IDENTITY_HEADER, to set an
       		identity (i=) for signing based on the value found in a
		particular header.  Requested by Florian Sager. 
	Fix inside dkimf_libstatus() to return extended status code 
		4.7.5 if temp-fail is due to key retrieval failure.
	Fix bug #SF2831720: Use new dynamic signature header generation
		code (see below).
	Add new exception handling code "keynotfound" (short form "key")
		and "On-KeyNotFound", which handles keys not found in DNS
		separately from other DNS errors.
	Fixes inside dkimf_libstatus() so that SMTP reply codes get set
		properly for temp-fails.
	Remove _FFR_COMMAIZE.
	LIBOPENDKIM: Add dkim_getsighdr_d(), a dynamic-length version of
		dkim_getsighdr().  The old function interface remains
		unchanged for backward compatibility.
	LIBOPENDKIM: Add dkim_dstring_printf().
	LIBOPENDKIM: Fix bug #SF2839858: Change "default_senderheaders" list
		to include only "from", per RFC5617.  Problem noted

opendkim/opendkim-lua.c
/*
**  Copyright (c) 2009, The OpenDKIM Project.  All rights reserved.
**
**  $Id: opendkim-lua.c,v 1.1.2.42 2009/12/01 18:29:38 cm-msk Exp $
*/

#ifndef lint
static char opendkim_lua_c_id[] = "@(#)$Id: opendkim-lua.c,v 1.1.2.42 2009/12/01 18:29:38 cm-msk Exp $";
#endif /* !lint */

#ifdef _FFR_LUA

/* system includes */
#include <sys/types.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <assert.h>

/* Lua includes */
#include <lua.h>
#include <lualib.h>

/* libopendkim includes */
#include <dkim.h>

/* opendkim includes */
#include "opendkim-lua.h"
#include "opendkim-db.h"
#include "opendkim.h"

/* local data types */
struct dkimf_lua_io
{
	_Bool		lua_io_done;
	const char *	lua_io_script;
};

/*
**  DKIMF_LUA_READER -- "read" a script and make it available to Lua
**
**  Parameters:
**  	l -- Lua state
**  	data -- pointer to a Lua I/O structure
**  	size -- size (returned)
**
**  Return value:
**  	Pointer to the data.
*/

static const char *
dkimf_lua_reader(lua_State *l, void *data, size_t *size)
{
	struct dkimf_lua_io *io;

	assert(l != NULL);
	assert(data != NULL);
	assert(size != NULL);

	io = (struct dkimf_lua_io *) data;

	if (io->lua_io_done)
	{
		*size = 0;
		return NULL;
	}
	else
	{
		io->lua_io_done = TRUE;
		*size = strlen(io->lua_io_script);
		return io->lua_io_script;
	}
}

/*
**  DKIMF_LUA_ALLOC -- allocate memory
**
**  Parameters:
**  	ud -- context (not used)
**  	ptr -- pointer (for realloc())
**  	osize -- old size
**  	nsize --  new size
**
**  Return value:
**  	Allocated memory, or NULL on failure.
*/

static void *
dkimf_lua_alloc(void *ud, void *ptr, size_t osize, size_t nsize)
{
	if (nsize == 0 && osize != 0)
	{
		free(ptr);
		return NULL;
	}
	else if (nsize != 0 && osize == 0)
	{
		return malloc(nsize);
	}
	else
	{
		return realloc(ptr, nsize);
	}
}

/*
**  DKIMF_LUA_SETUP_HOOK -- hook to Lua for handling a message during setup
**
**  Parameters:
**  	ctx -- session context, for making calls back to opendkim.c
**  	script -- script to run
**  	name -- name of the script (for logging)
**  	lres -- Lua result structure
**
**  Return value:
**  	2 -- processing error
**  	1 -- script contains a syntax error
**  	0 -- success
**  	-1 -- memory allocation failure
**
**  Side effects:
**  	lres may be modified to relay the script's signing requests, i.e.
**  	which key/selector(s) to use, whether to use "l=", etc.
** 
**  Notes:
**  	Called by mlfi_eoh() so it can decide what signature(s) to apply.
**
**  	Will require the ability to access databases, i.e. the
**  	dkimf_db_*() functions and the "conf" handle that contains
**  	references to available databases.  opendkim.c will need to export
**  	some functions for getting DB handles for this purpose.
*/

int
dkimf_lua_setup_hook(void *ctx, const char *script, const char *name,
                     struct dkimf_lua_script_result *lres)
{
	int c;
	int status;
	lua_State *l = NULL;
	struct dkimf_lua_io io;

	assert(script != NULL);
	assert(lres != NULL);

	io.lua_io_done = FALSE;
	io.lua_io_script = script;

	l = lua_newstate(dkimf_lua_alloc, NULL);
	if (l == NULL)
		return -1;

	luaL_openlibs(l);

	/*
	**  Register functions.
	**
	**  XXX -- turn this into a Lua library?
	*/

	/* request From domain */
	lua_register(l, "odkim_get_fromdomain", dkimf_xs_fromdomain);

	/* request source hostname */
	lua_register(l, "odkim_get_clienthost", dkimf_xs_clienthost);

	/* request a signature (domain, selector) */
	lua_register(l, "odkim_sign", dkimf_xs_requestsig);

	/* retrieve header/value */
	lua_register(l, "odkim_get_header", dkimf_xs_getheader);

	/* pass source IP to dkimf_checkip() */
	lua_register(l, "odkim_internal_ip", dkimf_xs_internalip);

	/* do POPAUTH check */
	lua_register(l, "odkim_check_popauth", dkimf_xs_popauth);

	/* request DB handle */
	/* XXX -- allow creation of arbitrary DB handles? */
	lua_register(l, "odkim_get_dbhandle", dkimf_xs_dbhandle);
	lua_pushnumber(l, DB_DOMAINS);
	lua_setglobal(l, "DB_DOMAINS");
	lua_pushnumber(l, DB_THIRDPARTY);
	lua_setglobal(l, "DB_THIRDPARTY");
	lua_pushnumber(l, DB_DONTSIGNTO);
	lua_setglobal(l, "DB_DONTSIGNTO");
	lua_pushnumber(l, DB_MTAS);
	lua_setglobal(l, "DB_MTAS");
	lua_pushnumber(l, DB_MACROS);
	lua_setglobal(l, "DB_MACROS");

	/* get number of envelope recipients */
	lua_register(l, "odkim_rcpt_count", dkimf_xs_rcptcount);

	/* get a specific envelope recipient */
	lua_register(l, "odkim_get_rcpt", dkimf_xs_rcpt);

	/* test DB for membership */
	lua_register(l, "odkim_db_check", dkimf_xs_dbquery);

	/* request an "l=" tag on new signatures */
	lua_register(l, "odkim_use_ltag", dkimf_xs_setpartial);

	/* request that the message be sent through verification */
	lua_register(l, "odkim_verify", dkimf_xs_verify);

	/* retrieve an MTA symbol */
	lua_register(l, "odkim_get_mtasymbol", dkimf_xs_getsymval);

	/* set up for re-signing */
	lua_register(l, "odkim_resign", dkimf_xs_resign);

	lua_pushlightuserdata(l, ctx);
	lua_setglobal(l, "ctx");

	switch (lua_load(l, dkimf_lua_reader, (void *) &io, name))
	{
	  case 0:
		break;

	  case LUA_ERRSYNTAX:
		if (lua_isstring(l, 1))
			lres->lrs_error = strdup(lua_tostring(l, 1));
		lua_close(l);
		return 1;

	  case LUA_ERRMEM:
		if (lua_isstring(l, 1))
			lres->lrs_error = strdup(lua_tostring(l, 1));
		lua_close(l);
		return -1;

	  default:
		assert(0);
	}

	status = lua_pcall(l, 0, LUA_MULTRET, 0);
	if (lua_isstring(l, 1))
		lres->lrs_error = strdup(lua_tostring(l, 1));

	lua_close(l);

	return (status == 0 ? 0 : 2);
}

/*
**  DKIMF_LUA_SCREEN_HOOK -- hook to Lua for handling a message after the
**                           verifying handle is established and all headers
**                           have been fed to it
**
**  Parameters:
**  	ctx -- session context, for making calls back to opendkim.c
**  	script -- script to run
**  	name -- name of the script (for logging)
**  	lres -- Lua result structure
**
**  Return value:
**  	0 -- success
**  	-1 -- failure
**
**  Notes:
**  	Called by mlfi_eom() so it can decide whether or not the message
**  	is acceptable.
**
**  	Will require the ability to access databases, i.e. the
**  	dkimf_db_*() functions and the "conf" handle that contains
**  	references to available databases.  opendkim.c will need to export
**  	some functions for getting DB handles for this purpose.
*/

int
dkimf_lua_screen_hook(void *ctx, const char *script,
                      const char *name, struct dkimf_lua_script_result *lres)
{
	int status;
	struct dkimf_lua_io io;
	lua_State *l = NULL;

	assert(script != NULL);
	assert(lres != NULL);

	io.lua_io_done = FALSE;
	io.lua_io_script = script;

	l = lua_newstate(dkimf_lua_alloc, NULL);
	if (l == NULL)
		return -1;

	luaL_openlibs(l);

	/*
	**  Register functions.
	**
	**  XXX -- turn this into a Lua library?
	*/

	/* test DB for membership */
	lua_register(l, "odkim_db_check", dkimf_xs_dbquery);

	/* request From domain */
	lua_register(l, "odkim_get_fromdomain", dkimf_xs_fromdomain);

	/* request DB handle */
	/* XXX -- allow creation of arbitrary DB handles? */
	lua_register(l, "odkim_get_dbhandle", dkimf_xs_dbhandle);
	lua_pushnumber(l, DB_DOMAINS);
	lua_setglobal(l, "DB_DOMAINS");
	lua_pushnumber(l, DB_THIRDPARTY);
	lua_setglobal(l, "DB_THIRDPARTY");
	lua_pushnumber(l, DB_DONTSIGNTO);
	lua_setglobal(l, "DB_DONTSIGNTO");
	lua_pushnumber(l, DB_MTAS);
	lua_setglobal(l, "DB_MTAS");
	lua_pushnumber(l, DB_MACROS);
	lua_setglobal(l, "DB_MACROS");

	/* retrieve header/value */
	lua_register(l, "odkim_get_header", dkimf_xs_getheader);

	/* get a specific envelope recipient */
	lua_register(l, "odkim_get_rcpt", dkimf_xs_rcpt);

	/* get number of envelope recipients */
	lua_register(l, "odkim_rcpt_count", dkimf_xs_rcptcount);

	/* retrieve number of signatures */
	lua_register(l, "odkim_get_sigcount", dkimf_xs_getsigcount);

	/* retrieve a signature handle */
	lua_register(l, "odkim_get_sighandle", dkimf_xs_getsighandle);

	/* retrieve a signature's domain */
	lua_register(l, "odkim_sig_getdomain", dkimf_xs_getsigdomain);

	/* retrieve a signature's identity */
	lua_register(l, "odkim_sig_getidentity", dkimf_xs_getsigidentity);

	/* ignore a signature and its result */
	lua_register(l, "odkim_sig_ignore", dkimf_xs_sigignore);

	lua_pushlightuserdata(l, ctx);
	lua_setglobal(l, "ctx");

	switch (lua_load(l, dkimf_lua_reader, (void *) &io, name))
	{
	  case 0:
		break;

	  case LUA_ERRSYNTAX:
		if (lua_isstring(l, 1))
			lres->lrs_error = strdup(lua_tostring(l, 1));
		lua_close(l);
		return 1;

	  case LUA_ERRMEM:
		if (lua_isstring(l, 1))
			lres->lrs_error = strdup(lua_tostring(l, 1));
		lua_close(l);
		return -1;

	  default:
		assert(0);
	}

	status = lua_pcall(l, 0, LUA_MULTRET, 0);
	if (lua_isstring(l, 1))
		lres->lrs_error = strdup(lua_tostring(l, 1));

	lua_close(l);

	return (status == 0 ? 0 : 2);
}

/*
**  DKIMF_LUA_FINAL_HOOK -- hook to Lua for handling a message after all
**                          signing and verifying has been done
**
**  Parameters:
**  	ctx -- session context, for making calls back to opendkim.c
**  	script -- script to run
**  	name -- name of the script (for logging)
**  	lres -- Lua result structure
**
**  Return value:
**  	0 -- success
**  	-1 -- failure
**
**  Notes:
**  	Called by mlfi_eom() so it can decide whether or not the message
**  	is acceptable.
**
**  	Will require the ability to access databases, i.e. the
**  	dkimf_db_*() functions and the "conf" handle that contains
**  	references to available databases.  opendkim.c will need to export
**  	some functions for getting DB handles for this purpose.
*/

int
dkimf_lua_final_hook(void *ctx, const char *script,
                     const char *name, struct dkimf_lua_script_result *lres)
{
	int status;
	struct dkimf_lua_io io;
	lua_State *l = NULL;

	assert(script != NULL);
	assert(lres != NULL);

	io.lua_io_done = FALSE;
	io.lua_io_script = script;

	l = lua_newstate(dkimf_lua_alloc, NULL);
	if (l == NULL)
		return -1;

	luaL_openlibs(l);

	/*
	**  Register functions.
	**
	**  XXX -- turn this into a Lua library?
	*/

	/* retrieve number of signatures */
	lua_register(l, "odkim_get_sigcount", dkimf_xs_getsigcount);

	/* retrieve a signature handle */
	lua_register(l, "odkim_get_sighandle", dkimf_xs_getsighandle);

	/* retrieve a signature's domain */
	lua_register(l, "odkim_sig_getdomain", dkimf_xs_getsigdomain);

	/* retrieve a signature's identity */
	lua_register(l, "odkim_sig_getidentity", dkimf_xs_getsigidentity);

	/* retrieve signature result */
	lua_register(l, "odkim_sig_result", dkimf_xs_sigresult);

	/* retrieve signature result (body hash check) */
	lua_register(l, "odkim_sig_bhresult", dkimf_xs_sigbhresult);

	/* size of body? */
	lua_register(l, "odkim_sig_bodylength", dkimf_xs_bodylength);

	/* canonicalized size? */
	lua_register(l, "odkim_sig_canonlength", dkimf_xs_canonlength);

	/* get a specific envelope recipient */
	lua_register(l, "odkim_get_rcpt", dkimf_xs_rcpt);

	/* get number of envelope recipients */
	lua_register(l, "odkim_rcpt_count", dkimf_xs_rcptcount);

	/* add recipient */
	lua_register(l, "odkim_add_rcpt", dkimf_xs_addrcpt);

	/* delete recipient */
	lua_register(l, "odkim_delete_rcpt", dkimf_xs_delrcpt);

	/* get policy code */
	lua_pushnumber(l, DKIM_POLICY_UNKNOWN);
	lua_setglobal(l, "DKIM_POLICY_UNKNOWN");
	lua_pushnumber(l, DKIM_POLICY_ALL);
	lua_setglobal(l, "DKIM_POLICY_ALL");
	lua_pushnumber(l, DKIM_POLICY_DISCARDABLE);
	lua_setglobal(l, "DKIM_POLICY_DISCARDABLE");
	lua_register(l, "odkim_get_policy", dkimf_xs_getpolicy);

	/* get reputation */
	lua_register(l, "odkim_get_reputation", dkimf_xs_getreputation);

	/* get policy result */
	lua_pushnumber(l, DKIM_PRESULT_NONE);
	lua_setglobal(l, "DKIM_PRESULT_NONE");
	lua_pushnumber(l, DKIM_PRESULT_NXDOMAIN);
	lua_setglobal(l, "DKIM_PRESULT_NXDOMAIN");
	lua_pushnumber(l, DKIM_PRESULT_AUTHOR);
	lua_setglobal(l, "DKIM_PRESULT_AUTHOR");
	lua_register(l, "odkim_get_presult", dkimf_xs_getpresult);

	/* set SMTP reply */
	lua_register(l, "odkim_set_smtp_reply", dkimf_xs_setreply);

	/* quarantine */
	lua_register(l, "odkim_quarantine", dkimf_xs_quarantine);

	/* set result code */
	lua_pushnumber(l, SMFIS_TEMPFAIL);
	lua_setglobal(l, "SMFIS_TEMPFAIL");
	lua_pushnumber(l, SMFIS_DISCARD);
	lua_setglobal(l, "SMFIS_DISCARD");
	lua_pushnumber(l, SMFIS_REJECT);
	lua_setglobal(l, "SMFIS_REJECT");
	lua_register(l, "odkim_set_result", dkimf_xs_setresult);

	lua_pushlightuserdata(l, ctx);
	lua_setglobal(l, "ctx");

	switch (lua_load(l, dkimf_lua_reader, (void *) &io, name))
	{
	  case 0:
		break;

	  case LUA_ERRSYNTAX:
		if (lua_isstring(l, 1))
			lres->lrs_error = strdup(lua_tostring(l, 1));
		lua_close(l);
		return 1;

	  case LUA_ERRMEM:
		if (lua_isstring(l, 1))
			lres->lrs_error = strdup(lua_tostring(l, 1));
		lua_close(l);
		return -1;

	  default:
		assert(0);
	}

	status = lua_pcall(l, 0, LUA_MULTRET, 0);
	if (lua_isstring(l, 1))
		lres->lrs_error = strdup(lua_tostring(l, 1));

	lua_close(l);

	return (status == 0 ? 0 : 2);
}
#endif /* _FFR_LUA */
